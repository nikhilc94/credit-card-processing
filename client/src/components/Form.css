ul {
  background-color: rgb(255, 0, 0, 0.5);
  padding: 10px 30px;
  width: 60%;
}

li {
  word-break: break-all;
}

form {
  display: flex;
  flex-direction: column;
  margin-bottom: 35px;
}

label {
  display: block;
  margin-bottom: 10px;
}

input {
  border: 1px solid rgb(204, 204, 204);
  display: block;
  font-size: 17px;
  margin: 5px 0 10px;
  padding: 8px;
  width: 25%;
}

button {
  background-color: rgb(216, 216, 216);
  border: rgb(168, 168, 168);
  font-size: 15px;
  padding: 7px;
  width: 10%;
}

@media only screen and (max-width: 768px) {
  input {
    font-size: 11px;
    width: 90%;
  }

  button {
    width: 25%;
  }
}
