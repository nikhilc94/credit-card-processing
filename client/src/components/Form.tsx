import axios from "axios";
import React, { useState } from "react";

import "./Form.css";
import fetchCards from "../utils/fetchCards";
import clientValidation from "../utils/clientValidation";

interface FormProps {
  setCardDetails: Function;
}

interface Event {
  persist: Function;
  target: any;
  preventDefault: Function;
}

const Form: React.FC<FormProps> = ({ setCardDetails }) => {
  const defaultState = {
    name: "",
    cardNumber: "",
    cardLimit: ""
  };

  const defaultErrorState: string[] = [];

  const [input, setInputDetails] = useState(defaultState);
  const [errors, setError] = useState(defaultErrorState);

  const handleInputChange = (e: Event) => {
    e.persist();
    setInputDetails(prevState => ({
      ...prevState,
      [e.target.name]: e.target.value
    }));
  };

  const handleSubmit = async (e: Event) => {
    e.preventDefault();
    setError(defaultErrorState);
    const clientErrors = clientValidation(input);
    if (clientErrors.length > 0) {
      setError(clientErrors);
      return;
    }
    try {
      await axios.post("/api/card", input);
    } catch (e) {
      setError([e.response.data.error]);
      return;
    }
    setInputDetails(defaultState);
    fetchCards(setCardDetails);
  };

  return (
    <div className="form-container">
      {errors.length > 0 && (
        <ul>
          {errors.map(err => (
            <li key={err}>{err}</li>
          ))}
        </ul>
      )}
      <form onSubmit={handleSubmit}>
        <label>
          Name
          <input
            type="text"
            name="name"
            required
            value={input.name}
            onChange={handleInputChange}
          />
        </label>
        <label>
          Card number
          <input
            type="number"
            name="cardNumber"
            required
            value={input.cardNumber}
            onChange={handleInputChange}
          />
        </label>
        <label>
          Limit
          <input
            type="number"
            name="cardLimit"
            required
            value={input.cardLimit}
            onChange={handleInputChange}
          />
        </label>
        <button type="submit">Add</button>
      </form>
    </div>
  );
};

export default Form;
