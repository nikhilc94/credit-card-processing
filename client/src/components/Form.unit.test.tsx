import React from "react";
import { shallow } from "enzyme";
import { createSerializer } from "enzyme-to-json";

import Form from "./Form";

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

describe("Form component", () => {
  it("renders successfully", () => {
    const wrapper = shallow(<Form setCardDetails={() => null} />);
    expect(wrapper).toMatchSnapshot();
  });
});
