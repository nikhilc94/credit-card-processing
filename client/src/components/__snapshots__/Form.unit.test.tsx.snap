// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders successfully 1`] = `
<div
  className="form-container"
>
  <form
    onSubmit={[Function]}
  >
    <label>
      Name
      <input
        name="name"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </label>
    <label>
      Card number
      <input
        name="cardNumber"
        onChange={[Function]}
        required={true}
        type="number"
        value=""
      />
    </label>
    <label>
      Limit
      <input
        name="cardLimit"
        onChange={[Function]}
        required={true}
        type="number"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
